### ISSUE TITLE
Help creating a custom program

### ISSUE THREAD

**Issue by KF243** [bicarlsen/easy-biologic#4]
[try1.zip](https://github.com/bicarlsen/easy-biologic/files/7342259/try1.zip)

On the vanilla Python environment, my example.py results in showing this error message.
Please suggest how this problem is fixed.

**Comment by KF243** [comment]
Python is upgraded to version 3.9.6, and then the result changes to the following.

min() arg is an empty sequence

[try2.zip](https://github.com/bicarlsen/easy-biologic/files/7343046/try2.zip)

**Comment by KF243** [comment]
voc.csv and jv.csv files are created at "C:\Users\kf\data\" when the last error message was shown as follows, 
  File "C:\Users\kf\AppData\Local\Programs\Python\Python39\lib\site-packages\easy_biologic\base_programs.py", line 1700, in <dictcomp>
    ch: min( data, key = lambda d: d.power ) # power of interest is negative
ValueError: min() arg is an empty sequence
Is the "data" folder created at the correct path?

**Comment by KF243** [comment]
I re-installed easy-biologic on the anaconda python and could create jv.csv voc.csv and mpp.csv files in data folder at the top directory. I also confirmed the contents in these files. So I think the problem might be caused by the wrong installing of easy-biologic in my python PC.
At the next step, I will code to show the results as the graphs with matplotlib and change the details of method.

**Comment by KF243** [comment]
I can only find brief comments for methods in your Project description. Could you open the more detail explanation for them?

**Comment by KF243** [comment]
MPP may serve us Vmpp by OCV and JV scan if the cell wires are connected with no cell. However, an ocv.csv file with correct data, a jv.csv file without data and no mpp.csv file are created with the following messages, when the cell wires were connected with DC3 dummy cell (in standard 3 electrode connection and this connection was successfully operated with EC-Lab).
(biologicSP300) C:\Users\kfush>python C:\Users\kfush\data\try.py
Traceback (most recent call last):
  File "C:\Users\kfush\data\try.py", line 19, in <module>
    mpp.run( 'C:/Users/kfush/data' )
  File "C:\Users\kfush\anaconda3\envs\biologicSP300\lib\site-packages\easy_biologic\base_programs.py", line 1580, in run
    self.v_mpp = self._run_jv( self.voc, jv_loc, by_channel = by_channel, jv_params = jv_params ) # jv
  File "C:\Users\kfush\anaconda3\envs\biologicSP300\lib\site-packages\easy_biologic\base_programs.py", line 1699, in _run_jv
    mpp = {
  File "C:\Users\kfush\anaconda3\envs\biologicSP300\lib\site-packages\easy_biologic\base_programs.py", line 1700, in <dictcomp>
    ch: min( data, key = lambda d: d.power ) # power of interest is negative
ValueError: min() arg is an empty sequence    

Which kind of problem is causing in my JV scan?

**Comment by bicarlsen** [comment]
The issue is that for some reason the `JV_Scan` program that is built into the `MPP` program isn't collecting any data. Can you please try running a `JV_Scan` program on it's own, and see if that works.

**Comment by KF243** [comment]
The problem depends on the cell and not on the program. Isn't it caused by using Ultra Low Current booster (https://www.biologic.net/products/ultra-low-current-down-to-aa-resolution/)?

**Comment by KF243** [comment]
There are two boards named 1ch and 2ch in EC Lab (0 and 1 in example.py), which are connected via each booster with a dummy cell or no cell. In EC lab, 1ch showed an Ohmic slope during CV but 2ch gave current over load and failed CV. In the same connection, try30.py (line 15:    channels = [0] ) created jv.csv (very small current and no ohmic slope) for 1ch and try31.py (line 16:    channels = [1] ) 
created jv.csv (no data) for 2ch
[try3.zip](https://github.com/bicarlsen/easy-biologic/files/7361047/try3.zip)
.

**Comment by bicarlsen** [comment]
For channel 2 the failure is expected. The current overload will prevent the measurement from proceeding and the `JV_Scan` should fail by design. If the technique won't run using EC Lab it appears the issue is with your dummy cell, and not the `easy_biologic` program.

**Comment by KF243** [comment]
Alternately, I am trying the using JV_Scan directly as cbj.py.
It seems to run. But I do not know how the data is correctly measured.
How the data can be saved in csv file ?

[try4.zip](https://github.com/bicarlsen/easy-biologic/files/7363817/try4.zip)

**Comment by bicarlsen** [comment]
Ah, I understand. I am unsure as to why the measurement is failing. My suggestion would be to just run an `MPP_Tracking` program then. That way you can bypass teh `JV_Scan` which is giving you problems. Is that a viable solution for you?

**Comment by KF243** [comment]
Many thanks. Now, jv_scan data file is successfully created independent of cell connection.

For the next ca measurement, in which ch0 and ch1 are independently polarized at different voltages, eg. 0.1 and 0.2 V, respectively, how do I code it? Using the attached py, both chs seems to be polarized at 0.1 V.
[try5.zip](https://github.com/bicarlsen/easy-biologic/files/7369187/try5.zip)
.

**Comment by KF243** [comment]
According your example, I modified for JV_Scan as follows but it does not work.
jv = blp.JV_Scan(device = 
    bl,
    params = { 
        '1': { 'start': 0, 'end': 1, 'step': 0.01 },
        '2': { 'start': -1, 'end': 0, 'step': 0.05 } 
        },
    channels = [0, 1]
    )

Does not need the dictionary 'quote' for a key?
For two chs (0 and 1), how does the params relate to channels?
Is the params of CA {'voltages', 'durations', etc} instead of { 'start', 'end', 'step', etc} ?

**Comment by bicarlsen** [comment]
As I mentioned, I removed much of the code to make it a bit more readable, so errors were going to occur. Unfortunately without a better understanding of the specifics of your project it is hard for me to give much input.

The main idea though, is that you can subclass either the `BiologicProgram` or `CA` classes and add your control code in the `_run` method. I have included the full example code below for your reference.

[control_example-full.zip](https://github.com/bicarlsen/easy-biologic/files/7473136/control_example-full.zip)

**Comment by KF243** [comment]
Please find the attachment which is the specifics of my project. During the XY stepping motions, current data (now, it is dummy) are collected to image. Initially potentiostatic polarization for 0ch and 1ch needs to be started. Current data of either ch at each XY step needs to be measured to image without disturbing polarization.

[try12.zip](https://github.com/bicarlsen/easy-biologic/files/7480107/try12.zip)

**Comment by KF243** [comment]
Using "segment" provides instantaneous data.
Is it possible for my target to collect current data in a csv file and to combine with scanning program?
Nevertheless, how to control the providing number of instantaneous data is unclear. In the attached case, time=interval:0.1, durations:10 for 2ch writes a couple of lines 10 times. How do I increase the line number in the same durations. 

[try13.zip](https://github.com/bicarlsen/easy-biologic/files/7502338/try13.zip)

**Comment by bicarlsen** [comment]
I'm not sure I completely understand your questions, but I will do my best to answer them.

Yes, you can write the data out to a file as you wish. I suggest looking at the `BiologicProgram.save_data` method for an example. Be aware that in the `DataSegment` there is both processd and unprocessed data. For the data to make sense you must access the `.data` attribute of the segement.

You can not control the amount of data coming from the BioLogic device. Essentially it will take a measurement whenever it surpasses one of the thresholds set up in the program parameters. This data is stored in a buffer. When `Biologic._retrieve_data` is called it pulls all the data out of this buffer.

Attached is some pseudocode that may be useful for you. I have not tested it, and there are likely errors, but hopefully it gives you an idea of ho I would approach the program from what I understand of it.
[secm.zip](https://github.com/bicarlsen/easy-biologic/files/7511058/secm.zip)

**Comment by KF243** [comment]
I am overwhelmingly lacking in reference information.
How I can find BiologicProgram.save_data and Biologic._retrieve_data ?

Thanks. Compared with mpp_lamp_track.py, certainly, secm.py may be useful for me. But it is still difficult for me to understand. Only for SECM( blp.CA) operation (without PositionController() operation), how are voltages and params set?
secm1.py results in the followings.

C:\Users\kfush>python C:\Users\kfush\data\secm1.py
Traceback (most recent call last):
  File "C:\Users\kfush\data\secm1.py", line 184, in <module>
    secm.run( 'data' )
  File "C:\Users\kfush\AppData\Local\Programs\Python\Python39\lib\site-packages\easy_biologic\base_programs.py", line 545, in run
    steps = len( ch_params[ 'voltages' ] )
KeyError: 'voltages'

[try14.zip](https://github.com/bicarlsen/easy-biologic/files/7518426/try14.zip)

**Comment by bicarlsen** [comment]
The documentation for both `BiologicProgram.save_data` and `Biologic._retrieve_data` is included in their docstrings which you can see in the [`program` file](https://github.com/bicarlsen/easy-biologic/blob/main/easy_biologic/program.py).

You can update a program's parameters using the [`BiologicDevice.update_parameters` method](https://github.com/bicarlsen/easy-biologic/blob/main/easy_biologic/device.py#L290). An example of how to use this is seen in the [`CA.update_voltages` method](https://github.com/bicarlsen/easy-biologic/blob/main/easy_biologic/base_programs.py#L561).

The reason your error is occurring is becasue of the way `params` are being passed in. If `params` is a dictionary keyed by the channels, you can not pass the `channels` parameter as well. Only use the `channels` parameter if all the channels use the same parameters, in which case the parameters should be a dictionary of the parameters only, not including the channels.
e.g.
```python
# all channels with same parameters
secm = SECM(  
    bl,
    coordinates,
    position_controller,
    params = { 'time_interval': 0.1, 'voltages': [ 0.1 ], 'durations': [ 10 ] },
    channels = [ 0, 1 ]
)

# or
# parameters specified by channel
secm = SECM(  
    bl,
    coordinates,
    position_controller,
    params = {
        0: { 'time_interval': 0.1, 'voltages': [ 0.1 ], 'durations': [ 10 ] },
        1: { 'time_interval': 0.1, 'voltages': [ 0.2 ], 'durations': [ 10 ] }
    }
)
```

**Comment by KF243** [comment]
line184: what is the role of argument 'data' for secm.run()?
SECM12.py can run because voltage of -0.2V for 10s at 1ch is detected actually but go somewhere without error message. How do I check the status of the program? How do I show the measured current?

[try15.zip](https://github.com/bicarlsen/easy-biologic/files/7531560/try15.zip)

**Comment by bicarlsen** [comment]
'data' was put here accidentally. It should be removed.

You can retrieve the measured current either from the `BiologicProgram.data` attritute, or you could read it in from the file it is being saved to.

What sort of status do you want from the program?

**Comment by KF243** [comment]
Line 184 in secm12.py was just corrected. But it still show nothing and does not stop. 
I want to know the "status" or situation, in which the program is running. Specially, I want to know which line is running or which line is wrong. 

To retrieve the measured current using BiologicProgram.data attribute or reading from the file, do I need to define a new "class" like PositionController or SECM?

**Comment by bicarlsen** [comment]
To track which line is running you should use an IDE that allows you to insert break points. I believe VSCode, PyCharm, and DataSpell allow this functionality, and there may be others. You can also place print or logging statements in the function to output what part of the program is currently running.

You don't need to create a new class to retrieve the data. If you are within the program you can use `self.data`, or from outside you can do something like `secm = SECM(); data = secm.data`.

**Comment by KF243** [comment]
It took a little time to develop VSCode environment in the PC. VSCode shows some errors for secm.py as follows.
---
Exception has occurred: AttributeError
'SECM' object has no attribute 'channel'
  File "C:\Users\kfush\data\secm13.py", line 156, in _run
    self.update_voltages( voltages, channels )
  File "C:\Users\kfush\data\secm13.py", line 186, in <module>
    secm.run()
---
How do I modify to attribute 'channel' for update_voltages?
Is it OK secm.run() is in line 186?

[try16.zip](https://github.com/bicarlsen/easy-biologic/files/7559264/try16.zip)

**Comment by bicarlsen** [comment]
Ah, apologies for this. This seems to be an issue with the `CA` program, try using the `CALimit` program instead.
i.e.
```python
class SECM( blp.CALimit ):
  ...
```

**Comment by KF243** [comment]
Line 55 was modified to 'class SECM( blp.CALimit ):', but the error message does not change.

Exception has occurred: TypeError
an integer is required (got type NoneType)
  File "C:\Users\kfush\data\secm13.py", line 156, in _run
    self.update_voltages( voltages, channels )
  File "C:\Users\kfush\data\secm13.py", line 186, in <module>
    secm.run()

[try16b.zip](https://github.com/bicarlsen/easy-biologic/files/7574145/try16b.zip)

**Comment by bicarlsen** [comment]
In your code the line occurs on the line `self.update_voltages( voltages, channels )`. First, `channels` should have probably been `self.channels`. But more importantly `update_voltages()` does not take channels as a parameter. Please be sure to look at the documentation of the function for its correct use.

**Comment by KF243** [comment]
Modified secm13.py shows the following error messages.
Exception has occurred: TypeError
float() argument must be a string or a number, not 'dict'
  File "C:\Users\kfush\data\secm13.py", line 158, in _run
    self.update_voltages( voltages )
  File "C:\Users\kfush\data\secm13.py", line 188, in <module>
    secm.run()

The documentation may be from line 560 in base_programs.py. But which is the documentation for self.channels? 

[try16c.zip](https://github.com/bicarlsen/easy-biologic/files/7577620/try16c.zip)

**Comment by bicarlsen** [comment]
`self.channels` is docuemnted as part of `BiologicProgram`. `voltages` should be a list of voltages keyed by channel. e.g.
```python
voltages = { 0: [ -1, 0, 1 ], 1: [ 0, -1, 1 ] }
```

**Comment by KF243** [comment]
I seems to have no BiologicProgram.

C:\Users\kfush\AppData\Local\Programs\Python\Python39\Lib\site-packages\easy_biologic>dir
 ドライブ C のボリューム ラベルは OS です
 ボリューム シリアル番号は 7063-5E1C です
 C:\Users\kfush\AppData\Local\Programs\Python\Python39\Lib\site-packages\easy_biologic のディレクトリ
2021/11/20  17:50    <DIR>          .
2021/11/20  17:50    <DIR>          ..
2021/11/20  17:49            56,654 base_programs.py
2021/11/20  17:49             2,182 common.py
2021/11/20  17:49            22,507 device.py
2021/11/20  17:49               829 find_devices.py
2021/11/20  17:50    <DIR>          lib
2021/11/20  17:49            26,058 program.py
2021/11/20  17:50    <DIR>          techniques-5.35
2021/11/20  17:50    <DIR>          techniques-6.01
2021/11/22  10:20    <DIR>          techniques-6.04
2021/11/20  17:49               352 __init__.py
2021/11/20  17:50    <DIR>          __pycache__
               6 個のファイル             108,582 バイト
               7 個のディレクトリ  859,412,389,888 バイトの空き領域

Anyway, the error message was changed as follows.
Exception has occurred: TypeError
an integer is required (got type NoneType)
  File "C:\Users\kfush\data\secm131.py", line 158, in _run
    self.update_voltages( voltages )
  File "C:\Users\kfush\data\secm131.py", line 188, in <module>
    secm.run()

[try16c.zip](https://github.com/bicarlsen/easy-biologic/files/7579246/try16c.zip)

**Comment by bicarlsen** [comment]
`BiologicProgram` is defined in the `program.py` file.

Unfortunately I'm not sure what is causing the new error. You may have to debug it yourself to find what is causing the issue.

**Comment by KF243** [comment]
In program.py, do I need to use self._channels and self.channels separately? How do I distinguish them?

**Comment by bicarlsen** [comment]
`BiologicPrograms._channels` is automatically set when the program is initialized so it is recommeneded to not modify it. `BiologicProgram.channels` returns the automatically configured channels, and is probably what you want to use.

**Comment by KF243** [comment]
param voltages should be "Dictionary of voltages list keyed by channel". "voltages = { 0: [ 0.1], 1: [-0.2] }" is not OK?
Which kind of integer is required to fix "voltages"?

[try17.zip](https://github.com/bicarlsen/easy-biologic/files/7624267/try17.zip)

**Comment by bicarlsen** [comment]
I don't think the attached program is the correct one.

**Comment by KF243** [comment]
"set_channel_configuration" for 2 chs are added.
But "set_channel_configuration" in line156 shows TypeError. You told a list of voltages keyed by channel. e.g. voltages = { 0: [ -1, 0, 1 ], 1: [ 0, -1, 1 ] }. How do I correct it?

[try18.zip](https://github.com/bicarlsen/easy-biologic/files/7648348/try18.zip)

**Comment by bicarlsen** [comment]
This is not an error with the parameters you are passing.

The error occurring is 
```python
File "C:\Users\kfush\AppData\Local\Programs\Python\Python39\lib\site-packages\easy_biologic\lib\ec_lib.py", line 988, in update_parameters
    idn = c.c_int32( idn )
TypeError: an integer is required (got type NoneType)
```
Meaning that the device is not connected when you are trying to update the parameters. My guess is this is occurring because on line 148 of `secm14.py` you call `self._disconnect()`, then try to update the voltages on line 157.

**Comment by KF243** [comment]
Version 0.3.2 was installed.

Since "asyncio.sleep()" at line 161 resulted in RuntimeWarning, the "asyncio.sleep()" is temporary not effective in secm15.py.
secm15.py seems to run but not show the CA data at each coordinate. How is it modified?

PS C:\Users\kfush\data> & C:/Users/kfush/AppData/Local/Programs/Python/Python39/python.exe c:/Users/kfush/data/secm15.py
<property object at 0x000001FDDFD68590>
[1 1 0]
[2 1 0]
[3 1 0]
[1 2 0]
[2 2 0]
[3 2 0]
[1 3 0]
[2 3 0]
[3 3 0]

[try19.zip](https://github.com/bicarlsen/easy-biologic/files/7664661/try19.zip)

**Comment by bicarlsen** [comment]
The reason you got an error when trying to run `asyncio.sleep()` in `SECM#_run` is because `#_run` is not an `async` method.

I cleaned up some of the code and logic from the program above. All the logical changes I made are in `SECM#_run`. Please see if it is a bit closer to what you want.

[secm15-edit.zip](https://github.com/bicarlsen/easy-biologic/files/7666572/secm15-edit.zip)

**Comment by KF243** [comment]
Only the first data ([110]_data.csv) contains CA data without position data as attached.

[Results_secm15-edit.zip](https://github.com/bicarlsen/easy-biologic/files/7667031/Results_secm15-edit.zip)

**Comment by bicarlsen** [comment]
I updated the program again. The functionality of `update_parameters` is a bit vague, as I'm not sure if the device "restarts" the program running the program from the beginning of the newly updated parameters, or if it continues from it's timed position, ignoring any parameters that have already been passed, even if they are updated. So you may have to play with the values of the `voltages` and `durations` parameters.

Again, I have not tested this code, so you may have to fix some errors.

[secm15-edit.zip](https://github.com/bicarlsen/easy-biologic/files/7667847/secm15-edit.zip)

**Comment by KF243** [comment]
Modified program (secm15-edit21.py) obtained the attached file (data.csv), in which CA data of [1 1 0] coordinate without x,y,z data. 

[try20.zip](https://github.com/bicarlsen/easy-biologic/files/7672835/try20.zip)

**Comment by bicarlsen** [comment]
It seems that for some reason the `x`, `y`, and `z` variables are `None`. I suggest looking at `SECM#go_to_position` to see if `self.position` is being set correctly and at the `_field_values` function defined in `SECM#__init__` to see why they aren't being set.

Also `self.position_measurement_time` should probably just be the sum of the duration from one of the channels, not both. As the channels run in parallel, summing their durations together will give double the actual run time.

**Comment by KF243** [comment]
The latest program (secm15-edit22.py) can move the position actually at each coordinate, because the movements are confirmed with naked eye. It seems work as a scanner. However, the saved file (data.csv) contains no header and x, y, z data for the first coordinate, and no data for the 2nd and more coordinates as attached. How are x, y, z data appended to CA data?

[try21.zip](https://github.com/bicarlsen/easy-biologic/files/7690302/try21.zip)

**Comment by bicarlsen** [comment]
The reason no header is being output is because `append = True` in the `save_data` call. You can add the header by creating a boolean called something like `already_run` and setting it to `False` outside the `for coordinate in self.coordinates` loop. Then after the `save_data` call, set it to `True`.
e.g.
```python
already_run = False
for coodinate in self.coordinates:
  # ...
  self.save_data( 'data.csv', append = already_run )
  already_run = True
```

I'm not sure why the coordinates are not bein saved in the data. I suggest adding some print statements in the `_field_values` function to try debugging. I would start by printing out `self` and `self.position` to see if those are being set correctly. If they are, then I would try making sure the tuple being returned is correct.

**Comment by KF243** [comment]
Now "data.csv" includes the header.
"_field_values" function seems is carried out once before the for "coordinate in self.coordinates: "loop as the printout.
How "_field_values" function is included in the loop? In the loop, "datum" and "segment" seems to be unknown.

[try22.zip](https://github.com/bicarlsen/easy-biologic/files/7696287/try22.zip)

**Comment by bicarlsen** [comment]
`_field_values` caluclates the values to be written out as data from the `DataSegment`.

Try retrieving the `position` form a function. e.g.
```python
def _field_values( datum, segment ):
  # ...
  position = self.get_position()
  # ...

# ...
def get_position( self ):
  return self.position
```

**Comment by KF243** [comment]
Although it was difficult for me to retrieve the position as your suggestion, now "data.csv" obtained by "secm16.py" includes x, y, and z data for the 1st position. I want to add the following position from the 2nd. How do I retrieve "_field_values( datum, segmant)" as well as x, y, z data in for coordinate loop from line 198.

[try23.zip](https://github.com/bicarlsen/easy-biologic/files/7709859/try23.zip)

**Comment by bicarlsen** [comment]
`self._field_values` is function called whenever data segments are read from the device, and is passed each data segment.

The error you made in your script is that you reference `controlX` instead of something like `self.controlX` in `SECM#get_position` and `_field_values`, and `PositionController#get_xyz`.
e.g.
```python
class PositionController():
    def __init__( self, controlX, controlY, controlZ ):
        self.x = controlX
        self.y = controlY
        self.z = controlZ

    # ...

    def get_xyz( self ):
        xx = self.x.position
        yy = self.y.position
        zz = self.z.position
        return ( xx, yy, zz )


class SECM( blp.CALimit, PositionController ):
    def __init__( ... ):
        self.position_controller = position_controller
        # ...
        def _field_values( datum, segment ):
            # ...
            xx, yy, zz = self.get_position()
    
    # ...
    def get_position( self ):
        return self.position_controller.get_xyz()
```
You may also be able to call `self.position_controller.get_xyz` directly in `_field_values` and eliminate `SECM#get_position`, or you may have to modify `SECM#get_position` to be something like
```python
def get_position( self ):
    xx = self.position_controller.x.position
    yy = self.position_controller.y.position
    zz = self.positon_controller.z.position

    return ( xx, yy, zz )
```
or you could take the coordinates from those passed in rather than from the position controller. 
e.g.
```python
def get_position( self ):
    return self.position
```

You may have to play around a bit with combinations of these.

**Comment by KF243** [comment]
"get_position" function is now in "_field_values" as SECM16-2.py.
However, return values of " _field_values" are not shown with position scanning and in data-SECM file (in which only data of the 1st position is recorded but no data of the 2nd and more).
How do I carry out "_field_values" or retrieve data for the following positions?

[try24.zip](https://github.com/bicarlsen/easy-biologic/files/7724311/try24.zip)

**Comment by bicarlsen** [comment]
My guess as to why the rest of the data isn't being collected is because of what I [brought up earlier](https://github.com/bicarlsen/easy-biologic/issues/4#issuecomment-987831365) in that when updating the voltages the program does not restart running from the beginning, so the program actually finishes after the first run. You may try playing with the `voltages` and `durations` params, changing them to something like
```
num_positions = soordinates.shape[ 0 ]
base_interval = 0.1
time_interval = base_interval * 0.5
durations =  [ base_interval ]* num_position
voltages = {
    0: [  0.1 ]* num_positions,
    1: [ -0.2  ]* num_positions
}
params = {
        0: {'time_interval': time_interval, 'voltages': [ 0.1 ], 'durations': durations },
        1: {'time_interval': time_interval, 'voltages': [ -0.2 ], 'durations': durations }
}
```
or in `#_run`
```python
def _run( self, technique, params, fields = None, interval = base_interval, retrieve_data = True ):
    # ... 

    for coordinate in self.coordinates:   
        self.go_to_position( coordinate )
        xx, yy, zz = self.get_position()
            
        # set voltages here
        self.update_voltages(
            self.params[ 0 ][ 'voltages' ],
            durations = self.params[0][ 'durations' ]
        )

        # perhaps change the sleep time to 0 to see if new positions are recorded
        time.sleep( self.position_measurement_time )
      # ...
```

At this point you may have to play around with some things to try to get the basic functionality working, then fine tune from there.

You can see how to manually retrieve data in the [`BiologicProgram#_retrieve_data_segment`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/program.py#L499) and [`BiologicProgram#_retrieve_data_segments`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/program.py#L539) methods.

**Comment by KF243** [comment]
Unnecessary commands were removed and the coding was arranged. Some "voltages" and "durations" params were tried. But I am still missing the rest of the data as attached.  It seems not to be carried out  line 144: asyncio.run( self._retrieve_data( interval ) ). 


[try25.zip](https://github.com/bicarlsen/easy-biologic/files/7733880/try25.zip)

**Comment by bicarlsen** [comment]
`asyncio.run( self._retrieve_data( interval * 5 ) )` is being run, otherwise no data would be retrieved. I think the issue has to do with the timing. I suggest going back to a version of the program where data was seemingly being retrieved for multiple positions, but the coordinates were not being saved, and step by step modifying it to save the coordinate positions as happens now.

**Comment by KF243** [comment]
Previously I think I failed to retrieve the rest data. 
The modified version excludes position scanning, but at line 124: def _run(), "fields" and "retrieve_data" are not accessed (which were also shown in the last coding: SECM16-3.py. May I restart the modification from this version or more primitive one? Anyway, no rest data data are saved.

[try26.zip](https://github.com/bicarlsen/easy-biologic/files/7734441/try26.zip)

**Comment by bicarlsen** [comment]
I'm sorry, I don't quite understand the issue you're having, or question you're asking. Could you try to clarfiy or provide a more indepth example?

**Comment by KF243** [comment]
I think that secm16-3retry.py in try26.zip, which has no relation with position scanning. However it cannot retrieve the 2nd and following data including CA output as data_SECM.csv. Why the 2nd data are ignored to save in the file.

**Comment by bicarlsen** [comment]
It appears that the program is running correctly. You commented out all the code that would iterate over the coordinates (lines 144 -167), and you only set one voltage in the parameters (line 44 - 47), so the additional durations are ignored.

**Comment by KF243** [comment]
Line 144-167 was modified. How are not ignored additional durations ?

[try27.zip](https://github.com/bicarlsen/easy-biologic/files/7787357/try27.zip)

**Comment by bicarlsen** [comment]
The `voltages` and `durations` parameters should have the same length. If they are not, the shorter one will cause the program to terminate once it is complete.

I think, however, you have a fundamental misunderstanding of how to use the program. e.g. Your calls to `update_voltages` are incorrect. Please look at the documentation or the `base_progrmas.py` file for examples of how to make these sorts of calls.

**Comment by KF243** [comment]
Smaller length (<= 8) of voltages and durations parameters are accepted but larger length (>=9) are not operated as attached. 
Why the update number is limited?

[try28.zip](https://github.com/bicarlsen/easy-biologic/files/7828350/try28.zip)

**Comment by bicarlsen** [comment]
I'm not 100% sure why this is occuring, but you are using the `CALimit#update_voltage` method incorrectly. To update the different channels, you should pass in a dictionary keyed by channel to each of the parameters. If a list is passed, then all channels are updated.

The error is
```
[-102] ERR_INSTR_TOOMANYDATA: too many data to transfer from the instrument (device error).
```
which isn't incredibly clear. Perhaps it would be worth contacting Biologic and asking them. I've worked with them before and they are quite responsive.

**Comment by KF243** [comment]
A dictionary keyed params for update_voltages was reconsidered, resulting in success to voltage update and save CALimit data. I am re-cording to combine stage scan and CALimit measurement. However, position data seem to be lost as csv-file, probably due to failure of synchronizing. How do I retrieve them?


[try30.zip](https://github.com/bicarlsen/easy-biologic/files/7832605/try30.zip)

**Comment by bicarlsen** [comment]
Glad to hear it's working better.

My guess as to why the position isn't working is because `SECM` both inherets `PositionController` and has a `PositionController` instance as an attribute (`SECM.position_controller`). Then, in the code you use the two interchangeably, which they shouldn't be.
E.g.
```python
# [line 130 in SECM#get_position]
return self.position_controller.get_xyz()  # gets the position of the position controller passed to the SECM instance.

# [line 155 in SEMC#_run]
self.go_to( coordinate )  # tells the PositionController represented by the SECM instance itself to go to the coordinate. 
```
So you can see you have two seperate instances of a `PositionController`. When you call `self.go_to` that is setting the position of the `PositionController` represented by the `SECM` instance, but has not changed the position of `SECM.position_controller`. So when you call `SECM#get_position` in `_field_values` it is returning the position of `SECM.position_ccontroller` which has not been changed.

My suggestion is to remove the inheritance of `PositionController` from `SECM` and only use the `PositionController` that is passed in as an argumenr (i.e. `self.position_controller`).

**Comment by KF243** [comment]
Position scanning might start after CALimit data reading as printout.txt.

In secm171.py, the modification for remove of the inheritance of PositionController from SECM and use of the PositionController is correct? 

[try31.zip](https://github.com/bicarlsen/easy-biologic/files/7837240/try31.zip)

**Comment by bicarlsen** [comment]
To remove inheritence change `class SECM( blp.CALimit, PositionController )` to `class SECM( blp.CALimit )`.

**Comment by KF243** [comment]
The inheritence of PositionController was removed from class SECM( blp.CALimit ). 
However, it still prints out datum (105) and then scan the position (from 128). How synchronize the scan to CALimit operation? Isn't it related to usage of _field_values( datum, segment )?

[try32.zip](https://github.com/bicarlsen/easy-biologic/files/7845365/try32.zip)

**Comment by bicarlsen** [comment]
It appears you have the experiment running for quite a short time. I suggest trying to increase the run time to see if that helps resolve the issue, especially in the beginning when everything is getting set up. I can also suggest removing teh `update_voltages` call from `SECM#_run` temporarily to see if you can take measurements and move the position controller at the same time. Once you have this simplified version of the program running, add the `update_voltages` call back in to coordinate the voltages and position.

**Comment by KF243** [comment]
I extended durations to 1 s, non-synchronization is not resolved. 
(Firstly, CALimit data are appeared and then stages are driven. The CALimit operation and stage drives are separately carried out.)
Furthermore,  a part of data for ch1 from 7.1 s are not saved.


[try33.zip](https://github.com/bicarlsen/easy-biologic/files/7851325/try33.zip)

**Comment by bicarlsen** [comment]
The reason all the measurements are happening before the positions change is because `asyncio#run` (line 126) is a blocking call. To allow both to run at the same time you can move the `for coordinate in self.coordinates` loop (lines 129 - 148) into an [`async coroutine`](https://docs.python.org/3/library/asyncio-task.html), let's call it `async def move_positions()`. Then create another function to [`gather`](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) this and the `self#_retrieve_data` calls, lets call it `async def main()`.
```python
async def main():
    await asyncio.gather( [ self._retrieve_data( interval ), move_positions() ] )
```
Finally, instead of calling `asyncio.run( self._retrieve_data( interval )  )` call `asyncio.run( main() )` (line 126).

I don't know why the data did not get saved for channel 1 after a certain time.

**Comment by KF243** [comment]
await asyncio.gather( [ blp.CALimit._retrieve_data( base_interval ), move_positions() ] )
seems not to be operated as the attached output. 
Locations of async def move_position & async def main() are correct?

[try34.zip](https://github.com/bicarlsen/easy-biologic/files/7860224/try34.zip)

**Comment by bicarlsen** [comment]
`async def main` should be a method of `SECM`.

**Comment by bicarlsen** [comment]
You'll also likely want to make [`Tasks`](https://docs.python.org/3.7/library/asyncio-task.html#creating-tasks) out of your coroutines. You can see [this StackOverflow](https://stackoverflow.com/questions/57966935/asyncio-task-vs-coroutine) thread for an explanation.

**Comment by KF243** [comment]
In the secm175.py, async def main and async def move_positions are methods of SECM.  List was removed from gather.  However, secm.run() seems to be a trouble maker as the printout.

[try35.zip](https://github.com/bicarlsen/easy-biologic/files/7861929/try35.zip)

**Comment by bicarlsen** [comment]
`main` shoul be a normal instance method not a class method. You should pass `self` to it, and run both the coroutines as instance methods.
E.g.
```python
async def main( self ):
        await asyncio.gather(
              self._retrieve_data( base_interval ), 
              self.move_positions() 
        )
...
asyncio.run( self.main() )
```

I also suggest not using `base_interval` in `_retrive_data` as it breaks encapsulation of the class. You should either pass that value in to `SECM`, calculate it from another property of the instance, or use a static value.

**Comment by KF243** [comment]
The stage is scanned and then CAlimit data for 10s (= duration) seem to be saved in csv.file. In csv.file, parts of voltages (0V at 0s, 0.1V at 0.5-3s, 0.2V at 3.5-4.5s, 0.3V at 5-7s, 0.4V at 7.5-8.5s, 0.5V at 9-10s for ch0) are irregularly recorded as well as the final position (x=200, y=200, z=100), and CAlimit data of 0.6-0.9V for ch0 are missed. First I want to start the position scanning and CALimit retrieving simultaneously.

[try36.zip](https://github.com/bicarlsen/easy-biologic/files/7868010/try36.zip)

**Comment by bicarlsen** [comment]
when calling `SECM#save_data` you set the `append` argument to `append_run` which is always False, so the data file is being overwritten. 

We added the `already_run` variable, so the headers would be printed out during the first save when it is `False`, but needs to be set to `True` after the first save. i.e. Add the line `already_run = True` after the `self.save_data()` call in `SECM#move_positions`.

I'm not sure why not all the data is being recorded, but it may be worth calculating the voltages and durations for all position initially, rather than updating them each time. 

For synchronization, you can read about the [`asyncion synchronization primatives`](https://docs.python.org/3/library/asyncio-sync.html). It's not immediately clear what th best way to synchronize is, though. Becase you can't reset the Biologic program at each position, it seems the only way to accomplish it would be to set a hold voltage, perhaps with an extended duration, at the end of each position measurement, then inspect the data when it is saved to see if it is in the holding period and move positions.
e.g. You want to measure at 0.1, 0.2, and 0.3 V, each for 1 second, and have a hold voltage at 0 V. this measurement should take place at 3 positions. Then you could do
```python
coordinates = [ 0, 1, 2 ]
voltages = [ 0.1, 0.2, 0.3, 0 ]* len( coordinates )
durations = [ 1, 1, 1, 10 ]* len( coordinates )
```
Then in `SECM#move_positions` you would check to see if the voltage was 0, and if so move to the next position.

**Comment by KF243** [comment]
'already_run = True' was added, but 'already_run' seems not to be referred  successfully.
[try37.zip](https://github.com/bicarlsen/easy-biologic/files/7879111/try37.zip)


While now, the desired operations for SECM are as follows,
1) Driving the position stages
2) Amperometry at constant potentials using two channel electrodes.
Remark for 1): x stage drives frequently (with a constant interval) but sometimes xyz stages also drive (taking different interval), meaning that stages do not drive at certain interval totally.
Remark for 2): at each x stage drive with a constant interval, current data should be recorded as well as stage position for current imaging. Only in tentative cordings, potentials (voltages) are changed to check the operation. 
On the other hand, CALimit or CA tends to collect data with the interval defined as params. So, I am wondering whether asyncio is available to retrieve current data with the intervals which is determined by stage drive. 

In chronoamperometry of EC-Lab, current data are displayed with some intervals during continuous polarization. Is it also possible to display (or print out) data using easy-biologic?

**Comment by bicarlsen** [comment]
In terms of `already_run` the error speaks for itself. Because you are using `already_run` in several locations throughout the code, I would probaby create an instance variable for it, e.g. `self.already_run`. 

That being said, this is a basic coding concept and you should expend some effort on fixing these errors yourself. I am happy to help you with coding the program, but you should not rely on me to fix every small bug that comes up in the program.

I don't understand your remarks enough to make any comment on them.

`easy_biologic` does not have any built in visualization tools. However you can easy add them by creating an external program that reads the data file being written to, or by creating a program that updates using the `BiologicProgram#on_data` callback.

**Comment by KF243** [comment]
I think that "self.already_run" works now. But line 127 of "for self.coordinate in coordinates: "(line 115-131) in move_positions() seems not to save the data. Data are collected and saved after the "for self.coordinate in coordinates: " or the last coordinate in "for self.coordinate in coordinates: ". 

[try38.zip](https://github.com/bicarlsen/easy-biologic/files/7887304/try38.zip)

**Comment by bicarlsen** [comment]
It appears that in `SECM#move_positions` you don't wait for any time at any of the positions. You need to use [`asyncio.sleep`](https://docs.python.org/3/library/asyncio-task.html#sleeping), not `time.sleep` in `async` functions, otherwise the entrie program will block.

**Comment by KF243** [comment]
Thanks every time. Addition of asyncio.sleep provides significant progress as attached files which include both position and CA data. 
However, some data are missed in saved file. Is it possible to be retrieved?
Furthermore I am wondering the difference between total saved time and controlling "params". Although I will try to minimize or diminish them, are they essential using asyncio.sleep or program.py?

[try39.zip](https://github.com/bicarlsen/easy-biologic/files/7896076/try39.zip)

**Comment by bicarlsen** [comment]
Glad to hear it's working better.

If you are referring to the blank lines in the data file, those are due to the asynchoronous nature of the program, but do not indicate any data is missing. Looking at the data, it appears everything is there.

I not quite sure what you mean by `total saved time` or `controlling "params"`.

**Comment by bicarlsen** [comment]
I should also mention, that I have written some utility functions for importing the data created from `easy-biologic` programs into `pandas`. It is part of the [`bric-analysis-libraries`](https://pypi.org/project/bric-analysis-libraries/) package, in the [`bric_analysis_libraries.jv.biologic_data_prep`](https://github.com/bicarlsen/analysis-libraries/blob/main/bric_analysis_libraries/jv/biologic_data_prep.py) module.

You may also be interested in the [`bric_analysis_libraries.jv.jv_data_prep`](https://github.com/bicarlsen/analysis-libraries/blob/main/bric_analysis_libraries/jv/jv_data_prep.py) and [`bric_analysis_libraries.jv.jv_analysis`](https://github.com/bicarlsen/analysis-libraries/blob/main/bric_analysis_libraries/jv/jv_analysis.py) modules.

**Comment by KF243** [comment]
I am changing some parameters at lines 35-42.
1) It is difficult to complete synchronize position and CA data. See the data. For example, x=100um at Time=1.5s but still Voltage=0.1V (should be 0.2V) for ch0.
Do I need additional trick for "asyncio.gather()"?
2) I take ca. 1.5s to change the position or update voltage. To decrease this period, I am wondering to change any parameters other than those of lines 35-42? 

[try40.zip](https://github.com/bicarlsen/easy-biologic/files/7904073/try40.zip)

**Comment by bicarlsen** [comment]
For the position synchronization, I would try exchanging the `self.update_voltages` and `position_controller.go_to` calls, so the voltage is updated only after the new position is obtained. I believe this will also allow you to decrease the time between positions.

**Comment by KF243** [comment]
I exchanged the self.update_voltages and position_controller.go_to calls. In saved data, a parts of CA data seems not to be synchronized as at 3.156s, 4.428s, 8.416s, 9.726s, 12.260s, and 13.561s.

[try41.zip](https://github.com/bicarlsen/easy-biologic/files/7911725/try41.zip)

**Comment by bicarlsen** [comment]
I see two options to try:

1. Try flushing the data by calling [`self._retrieve_data_segments`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/program.py#L539) before calling `self.save_data` in `SECM#move_positions`.
2. Increase the sleep time in `SECM#move_positions` so the data from the current position is flushed before moving to the next position. [`easy_biologic.program.BiologicProgram#_retrieve_data`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/program.py#L558) take an `interval` argument that dictates how often data is read from the device, in seconds. It defaults to 1 second. If you wait at least this amount of time extra, the data should automatically flush.

**Comment by KF243** [comment]
Since any sleep times during measurement directly will affect a total SECM imaging time, finally I want to decrease it as possible as small.  
Anyways, now irregular synchronization is accomplished as followed. 
Most of position and CA data are changed by "for" loop (at 111). But the first voltage update seems to be ignored and then CA data at previous condition are saved at each position.

Can I set "interval" of "_retrieve_data" as line 128? 
 
[try42.zip](https://github.com/bicarlsen/easy-biologic/files/7917735/try42.zip)

**Comment by bicarlsen** [comment]
It appears on line 117, in `SECM#move_positions`, you're not actually calling `self._retrieve_data_segments` which could be causing some of the issues. Also, it should go after the `asyncio.sleep` call, not before (but still before the `self.save_data` call, and should be `await`ed.
i.e.
```python
# your code
self._retrieve_data_segments
await asyncio.sleep( ... )
self.save_data( ... )

# should be
await asyncio.sleep( ... )
await self._retrieve_data_segments()
self.save_data( ... )
```

Yes, it appears you are calling `self._retrieve_data` with the `interval` parameter correctly.

**Comment by KF243** [comment]
The self.update_voltages and position_controller.go_to calls are exchanged. But I do not find the suitable intervals (line 36-45) for synchronization of position to update voltage. There are several hundred milliseconds-differences as the attached.

[try43.zip](https://github.com/bicarlsen/easy-biologic/files/7956861/try43.zip)

**Comment by bicarlsen** [comment]
I believe this is a consequence of running the `position_controller.go_to` and `self.update_voltages` synchronously. i.e. After `position.go_to` finishes it will take some time for `self.update_voltages` to actually update the voltages, and this is the delay you are seeing.

If you need a higher level of precision then you can try making both of these call asynchronous so the voltages will update while the position controller is moving. I see two ways to accomplish this.

 1. I would start by making the position controller asynchronous using [`asyncio`](https://docs.python.org/3/library/asyncio.html). You can then set the position controller to change position, then call `self.update_voltages` wrapped in a task, and use [`asyncio.gather`](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) to wait for both to complete before continuing with the loop, or possibly using an [`asyncio` synchronization primative](https://docs.python.org/3/library/asyncio-sync.html).

 2. If that doesn't solve the issue you can also try replacing `self.update_voltages` with a manual call to [`easy_biologic.lib.eclib#update_parameters_async`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/lib/ec_lib.py#L1395), again waiting for both.

Note that the `async` version of `easy_biologic.lib.eclib` functions are not well tested, so you may encouter erros. I also can not ensure how much improvement this will give you.

**Comment by KF243** [comment]
After each position control, an interval for stabilizing electrode reaction is necessary (line 72). I am wondering why return value (xx, yy, zz) of position_controller.go_to prior to update_voltages can not be retrieved. Asynchronous of position controlling can arrange in order correctly?

**Comment by bicarlsen** [comment]
You could retrieve the previous position because you have all of them stored in the `coordinates` list, or you could simply save the current position before going to the new position. However, I don't think this solves the synchronicity issue. The Biologic takes measurements as it sees fits, all we can do is poll it to see if it has taken any data. So the Biologic can take measurements during the movement process and there is no way to tell exaclty when the measurement was taken.

Think about what the case would be if you performed the steps manually. You would go to a position, set your voltage, and take your measurements. You would then either hold your voltage as you move, set the new voltage, and then perform your measurements at the new position. This is exactly what we are seeing in the data. So the question becomes, what do you want the voltage to do while you are in the process of changing positions.

There are only three options to the order the voltage update and position change can take place:
1. **Position before voltage.** Do you want the position to change before updating the voltages. If so, then this is what we have.
2. **Voltage before position.** Do you want to update the voltage before changing positions, instead of after? If so you can just swap the order of the `position_controller.go_to` and `self.update_voltages` calls, and adjust other code as needed.
3. **Voltage and position change together.** Do you want the voltage to change as the controller is moving to the new position? Then you need to use asynchronous methods as described previously.

I don't understand what you mean by
> Asynchronous of position controlling can arrange in order correctly?

Also the `for` loop running over `coordinates` seems very poorly written. I believe what you want is
```python
# currently
for self.coordinate in coordinates:

# should be
for coordinate in self.coordinates:
```
with the variables in the loop changing accordingly.

**Comment by KF243** [comment]
The for loop running over coordinates were corrected.

"self.update_voltages" with various voltages for each channel are used to check the data synchronization. After the confirmation of the synchronization it will change to that  with fixed voltages. 

Although synchronicity of position and current data are needed in the saved file, polarization and current measurement are conducted after the position set (position_controller.go_to).  Find the attached pdf which shows diagrams for motion and time schedule in SECM operation.


[try44.zip](https://github.com/bicarlsen/easy-biologic/files/7967301/try44.zip)

**Comment by bicarlsen** [comment]
I'm not sure what question you have in your previous comment.

**Comment by KF243** [comment]
Please forget about "Asynchronous of position controlling can arrange in order correctly?" It is difficult to convey it well. However, I believe that the diagrams in the last pdf.file shows my desired timing of stage motion and current measurement very well.

**Comment by bicarlsen** [comment]
It does, but that still isn't a question. I can't help you unless you have a specific question.

**Comment by KF243** [comment]
During the for loop of self.pos = 0 to 1, difference of 0.762s was recorded. Similarly 0.816s fro 1 to 2, 2.977s from 2 to 3, 0.891s from 3 to 4, 0.827s from 4 to 5, 3.027s from 5 to 6, 1.042s from 6 to 7, 
and 0.819s from 7 to 8. Larger differences (ca. 3 s) of 2to3 and 5to6 are probably due to "position_controller.detour( coordinate )". I guess that "position_controller.detour( coordinate )" needs to be excluded from asyncio.gather()? How do I decrease or diminish 0.8s-interval after each position?

![SS_SECM184](https://user-images.githubusercontent.com/92414221/151745368-1c70c269-8fc5-4e1c-949a-60b381147835.png)

**Comment by bicarlsen** [comment]
As I described before, I think the best way to get better synchronization would be to make the `PositionController` asynchronous so the voltage and position update in parallel, rather than in series.

**Comment by KF243** [comment]
I think that secm1851.py controls the voltage and position updates being asynchronous in parallel. However, saved data show differences of time.
Now, does "async def move_positions( self, already_run = False ):" or "async def main( self ):" make the problem?

[try45.zip](https://github.com/bicarlsen/easy-biologic/files/7976539/try45.zip)

**Comment by bicarlsen** [comment]
You haven't made either of the actual method you need to call -- `SECM#update_voltages` or `PositionController#go_to` asynchronous. All you've done is add an `asyncio#sleep` call after them.

To make `SECM#update_voltages` async you can do as I described above by manually calling `easy_biologic.lib.eclib#update_parameters_async. To make `PositionController#go_to` async, though, you'll need to reprogram it in an asyncronous way. It seems this can be done fairly easily by wrapping the `optosigma.GSC01#sleep_until_stop` method in an asynchronous way.

**Comment by KF243** [comment]
How do I call "easy_biologic.lib.eclib#update_parameters_async" manually? "ecl.update_parameters_async()" requests me idn, ch, and technique which seem to be int in ec_lib, as well as params? The following makes no sense.

await ecl.update_parameters_async(
      idn = 16,
      ch = 2,
      technique = 102,
      params = {
            0: {'time_interval': time_interval, 'voltages': voltages0, 'durations': durations },
            1: {'time_interval': time_interval, 'voltages': voltages1, 'durations': durations }
       }
)

**Comment by bicarlsen** [comment]
I'll refer you to the code that currently calls these methods as an example:
[`easy_biologic.base_programs.CALimit#update_voltages`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/base_programs.py#L976), and 
[`easy_biologic.device.BiologicDevice#update_parameters`](https://github.com/bicarlsen/easy-biologic/blob/81dc00692a008111d367f0643f15b9096ea2a23c/easy_biologic/device.py#L338)

I would encouage you to browse through the source code for examples if you have trouble understanding how to implement something in the future.

**Comment by bicarlsen** [comment]
I will close this issue due to inactivity. If something else comes up, please feel free to open a new Issue.
